### ISSUE TITLE
RFC: Handling of (non-UTF-8) byte payloads in Vector and VRL

### ISSUE THREAD

**Issue by pablosichert** [vectordotdev/vector#11577]
When ingesting arbitrary bytes, components within the Vector topology currently may handle the payload in any of these ways:
- preserve the payload
- lossy conversion into a UTF-8 string
- report an error for invalid UTF-8 encoding

Meaning, some combination of sources, transforms, sinks and their decoding/encoding settings may be able to handle non-UTF-8 data, others may not. However, we are not explicit to which level we support this.

Another argument in this discussion is log processing on Windows where UTF-16 encoding is often used.

**Comment by jszwedko** [comment]
Related: https://github.com/vectordotdev/vector/issues/10571

**Comment by fpytloun** [comment]
Might be related to this as well: https://github.com/vectordotdev/vector/discussions/12131

Causing this error when decoding JSON with some unsupported characters:
```
function call error for \"parse_json\" at (20:49): unable to parse json: invalid unicode code point at line 1 column 8587
```

**Comment by kevinburkesegment** [comment]
Just FYI the Datadog source still uses strict JSON mode to decode incoming payloads, and will reject payloads that contain invalid Unicode code points. The datadog-agent does not properly handle invalid Unicode code points: https://github.com/DataDog/datadog-agent/issues/35038

A "lossy" option for input here would have been helpful.

**Comment by jszwedko** [comment]
> Just FYI the Datadog source still uses strict JSON mode to decode incoming payloads, and will reject payloads that contain invalid Unicode code points. The datadog-agent does not properly handle invalid Unicode code points: [DataDog/datadog-agent#35038](https://github.com/DataDog/datadog-agent/issues/35038)
> 
> A "lossy" option for input here would have been helpful.

Agreed, if the Datadog Agent is capable of sending invalid Unicode code points it seems reasonable for Vector's `datadog_agent` source to handle receiving them.

**Comment by pront** [comment]
I agree, having an option to drop non-UTF8 characters makes sense. I wanted to comment in case someone is motivated to make a PR.

**Comment by kevinburkesegment** [comment]
I'm not sure I'm going to have time to submit a patch (and get the Rust toolchain up and going on my machine) but I'd appreciate if someone did want to do that :)
